caffe2/core/storage.h

```diff
@@ -77,117 +80,135 @@ class CAFFE2_API StorageImpl : public c10::intrusive_ptr_target {
     return data_ptr_.get();
   }
 
-  DataPtr& data_ptr() {
+  at::DataPtr& data_ptr() {
     return data_ptr_;
   }
 
-  const DataPtr& data_ptr() const {
+  const at::DataPtr& data_ptr() const {
     return data_ptr_;
   }
 
-  void set_dtype(const DataType& data_type) {
+  // Returns the previous data_ptr
+  at::DataPtr set_data_ptr(at::DataPtr&& data_ptr) {
+    std::swap(data_ptr_, data_ptr);
+    return std::move(data_ptr);
+  };
+
+  void set_dtype(const TypeMeta& data_type) {
+    int64_t capacity = numel_ * data_type_.itemsize();
     data_type_ = data_type;
+    numel_ = capacity / data_type_.itemsize();
```

Comment:
I think there are still problems here. We should not have this method at all.

Consider the following sequence of events:
1. I allocate a tensor of `std::string`
2. On the storage, I call `set_dtype` to `int`

If I have a tensor of `std::string`, I **must** placement-delete the entries before I finish the conversion to, otherwise I will leak the data in question. So a raw call to `set_dtype` must have arranged for placement-delete to have happened; it's an INVARIANT that must be obeyed by the caller, otherwise you will leak.